bolt-modules/boltlib/lib/puppet/functions/run_task_with.rb

```diff
@@ -124,7 +124,9 @@ def run_task_with_with_description(task_name, targets, description, options = {}
       # If parameters are mismatched, create a failing result for the target that will later
       # be added to the ResultSet.
       unless pcp_only
-        params = task.parameter_defaults.merge(params)
+        # Set the default value for any params that have one and were not provided
+        # If a parameter was given an undef value and has a default specified, use the default
+        params = task.parameter_defaults.merge(params) { |_, default, passed| passed || default }
```

Comment:
I think if we move this line outside the `pcp_only` then we don't have to do the merge twice? That will result in our wrapping default values for sensitive parameters in `Sensitive` which we don't currently do. That seems more correct to me anyway though.

Are we modifying the hash yielded from the block now? That shouldn't commonly be a problem because you probably are constructing the hash specifically to return it to the block, but if you've built up a map of target -> params instead, then that variable will be modified after this function runs.

Reply:
It appears we don't modify the hash in any case. The only time we _could_ modify it would be with `pcp_only == true` (since we don't make a copy by merging it with the defaults). But if `pcp_only == true`, we make a stub task and don't have any parameters so we will never wrap anything in Sensitive in that case (therefore making no modifications).

That also means my suggestion to move this line outside the conditional won't work, since `task.parameter_defaults` will be `nil` in the `pcp_only` case.

I do think the other suggestion of filtering `params` to only those with non-nil values early would reduce some duplication though.